### DISCUSSION TITLE
Moving Wayland rendering from wlroots to qtile

### DISCUSSION THREAD

**Discussion by QBobWatson** [qtile/qtile#4553] [Show and tell]
I've been thinking about how to implement some compositor eye candy goodness in qtile/Wayland.  It seems to me that the Wayland renderer should be hackable / scriptable like the rest of qtile.  So I started out by extracting some rendering code and rewriting it in python: https://github.com/QBobWatson/qtile/commit/722ab043c301f1944c6770fab80fe16d5d6b0b11

In wlroots, the method `wlr_scene_output_commit()` is in charge of pushing the scene graph through the renderer.  This can actually be done much more easily in Python, since it's mostly bookkeeping, so that's what I did.  It seems to work: in limited testing, I can't tell the difference with the wlroots code.  But now we're free to implement our own renderer in python (using python's opengl bindings) to add opacity, blur, rounded borders, etc.

I had to add a (surprisingly) few bindings to pywlroots (https://github.com/QBobWatson/pywlroots/commit/ae9c83bd0a854858155df4fb6201bfcafaa1a57c) to make this work.  I have to access some fields marked private in `wlr_scene_buffer` and `wlr_scene_node` -- presumably the wlr_scene api is meant to be used with a compositor's custom rendering pipeline, but doing so seems to require access to those fields.  Anyway we'll see how robust it is.

Next I'm going to rewrite the wlroots GLES2 renderer in Python, which should be pretty easy to do since it only knows how to draw rectangles.  Then I'll start jazzing it up.

Any thoughts?

**Comment by Shinyzenith** [comment]
I love freestanding implementations but I'd just like to play devils advocate for a second. Maybe contributing to a common impl for eye candy that isn't qtile specific helps out the whole community in a more impactful way. This helps other compositors use the same stack and over all more collaborative effort.

Whatever the qtile maintainers decide is obviously final, all I'm saying is contributing to scenefx with blur improvements etc might be more worth it in the long run. Not to mention that fx renderer already is battle tested. Also it's less of a maintainence burden

**Reply by Shinyzenith** [reply]
I'm not trying to discourage your work just to be clear!

**Reply by jwijenbergh** [reply]
I was just going to comment the same thing. I love your enthusiasm and recent PRs @QBobWatson. I think that right now contributing to scenefx is more beneficial to the whole wayland community.

That being said, interesting to see what you can of course do in Python itself so if you want to play around with that approach be my guest! The cool thing about scenefx is that it needs relatively little changes inside of qtile. If there's certain things that will not be implemented in scenefx because they're not common enough then making a renderer just for qtile is our next option, but I don't see that being the case right now.

**Reply by QBobWatson** [reply]
Thank you for the comments.  I think scenefx is a great project, don't get me wrong.  Give me a couple days to put together a working example though so you can see what I have in mind.

**Reply by dannycolin** [reply]
I would also prefer to see a generic/common library for this. This is something that's bothering me with Wayland currently. We're somewhat recreating the same issue we have with the Web where each "engine" don't support the same set of features. This ain't going to end well.

**Reply by QBobWatson** [reply]
Like I said, I think scenefx is a great project.

**Comment by QBobWatson** [comment]
Ok I've got some working code now.  Renderer: https://github.com/QBobWatson/qtile/commit/78a7b2376267cf881e1f0d82b696b21bd5da7a61 with pywlroots changes (in case you want to try it): https://github.com/QBobWatson/pywlroots/tree/jdr

To start, supporting a custom renderer requires minimal changes to existing qtile code.  I added a config option `wl_renderer` that `Core` picks up; `Output` uses it for rendering if it's there, and `Window` uses it to create window borders if it's there.  So you just have to set `wl_renderer = MyAwesomeRenderer` in your config file.

The interesting code is `Renderer` in `renderer.py`.  This isn't a port of `wlr_scene_output_commit` anymore; now it's just a renderer written in Python.  It traverses the scene tree, examining each node to figure out how to draw it.  By default, it does the same thing as the wlroots renderer would do.  But, if the node is a window container tree, it runs `render_window_borders` to render its borders.  When it sees a buffer descended from a window, it applies a heuristic to decide whether it's the main window content, then runs `render_main_window_surface`.  Other kinds of logic can be implemented in `render_window_buffer` to run different methods for popups, static windows, layers, etc.

(Aside: I believe Wayland was created with the idea that the clients would render their own decorations.  This seems like a terrible idea to me.  In particular, it means the base protocol gives the compositor no way to understand what the various surfaces created by the client are for.  The xdg_shell protocol helps a bit, but even then the compositor has no idea what to do with the subsurfaces.  For instance, Firefox puts all of its content in a subsurface with the same dimensions as its xdg toplevel.  So which surface should get a custom opacity?  ...)

At the moment, `Renderer` can render rounded corners and rounded borders (although I don't anticipate major difficulties with blur, drop shadows, and the like, beyond deciding which surfaces to apply them to):

![screenshot](https://github.com/qtile/qtile/assets/7871917/dfc40130-af06-4e53-b635-6f94252c3883)

But here's the key point, and the reason I think this project is worthwhile: _you can make your own Renderer_ to do whatever you want!  Do you want the border to be a gradient instead of a solid color?  Just subclass `Renderer` and use your own shader.

```python
class MyRenderer(Renderer):
    gradient_border_src = '''
#ifdef GL_FRAGMENT_PRECISION_HIGH
precision highp float;
#else
precision mediump float;
#endif

varying vec2 v_pos;
uniform vec4 color1, color2;
uniform float corner;
uniform float borderWidth;
uniform vec2 boxSize;


void main() {
  float len;
  if(corner > 0.0) {
    vec2 boxPos = (v_pos - vec2(0.5)) * boxSize;
    len = length(max(abs(boxPos) - boxSize / 2.0 + vec2(corner + borderWidth), vec2(0.0)));
    if(len > corner + borderWidth + 1.0) discard;
    if(len < corner - 0.5) discard;
  }
  gl_FragColor = mix(color1, color2, (v_pos.x + v_pos.y) / 2.0);
  if(corner > 0.0)
    gl_FragColor *= min(min(corner + borderWidth + 1.0 - len, len - corner + 0.5), 1.0);
}
    '''

    def init_shaders(self):
        super().init_shaders()
        self.programs['border_gradient'] = Program(
            self.shaders['common.vert'], self.gradient_border_src,
        )
        self.rotater = RGBRotate(-50)

    def render_window_borders(self, render_node):
        render_output = render_node.render_output
        program = self.programs['border_gradient']

        render_region = self.render_border_setup(render_node, program)
        if not render_region:
            return

        win = render_node.win
        color = rgb(win.bordercolor[0])
        GL.glUniform4f(program.uniforms['color1'], *color)
        r, g, b = self.rotater.apply(color[0], color[1], color[2])
        GL.glUniform4f(program.uniforms['color2'], r, g, b, 1)

        for rect in render_region.rectangles_as_boxes():
            self.scissor_in_damage_coords(render_output, rect)
            GL.glDrawArrays(GL.GL_TRIANGLE_STRIP, 0, 4)

        GL.glDisableVertexAttribArray(program.attributes['pos'])

        render_region.fini()

wl_renderer = MyRenderer
```

![screenshot](https://github.com/qtile/qtile/assets/7871917/9dfeeae6-5a46-4c48-bf64-3780f5218d98)

Do you want your borders to fade out?  Override `create_borders` and write your own shader:

```python
class MyOtherRenderer(Renderer):
    fade_border_src = '''
#ifdef GL_FRAGMENT_PRECISION_HIGH
precision highp float;
#else
precision mediump float;
#endif

varying vec2 v_pos;
uniform vec4 color;
uniform vec2 boxSize;
uniform float fadeFull, fadeLen;

void main() {
  float boxPos = v_pos.x * boxSize.x;
  float alpha = clamp(1.0 + (fadeFull - boxPos) / fadeLen, 0.0, 1.0);
  gl_FragColor = color * alpha;
}
    '''

    def init_shaders(self):
        super().init_shaders()
        self.programs['border_fade'] = Program(
            self.shaders['common.vert'], self.fade_border_src,
        )

    def create_borders(self, win):
        tree = win.container
        borders = [node.as_rect for node in tree.children
                   if node.type == SceneNodeType.RECT]
        # We use 6 border rects (one for each side and two corner boxes).
        # We can re-use the old borders if there are 6 of them.
        if len(borders) != 6:
            # Clear the old borders
            for rect in borders:
                rect.node.destroy()
            borders = []
        corner_w = corner_h = getattr(win, 'bordercorner', 0)
        if corner_w > win.width:
            corner_w = win.width
        if corner_h > win.height:
            corner_h = win.height
        bw = win.borderwidth

        # x1, y1, x2, y2, relative to window top-left
        geometries = (
            (-bw, -bw, corner_w, corner_h),  # top left
            (win.width - corner_w, win.height - corner_h,
             win.width + bw, win.height + bw),  # bottom right
            (corner_w, -bw, win.width + bw, 0),  # top
            (-bw, corner_h, 0, win.height + bw),  # left
            (win.width, 0, win.width + bw, win.height - corner_h),  # right
            (0, win.height, win.width - corner_w, win.height + bw),  # bottom
        )

        color = rgb(win.bordercolor[0])
        if borders:
            for (x1, y1, x2, y2), rect in zip(geometries, borders):
                rect.set_color(color)
                rect.set_size(x2 - x1, y2 - y1)
                rect.node.set_position(x1 + bw, y1 + bw)
        else:
            for (x1, y1, x2, y2) in geometries:
                rect = SceneRect.create(tree, x2 - x1, y2 - y1, color)
                rect.node.set_position(x1 + bw, y1 + bw)
        if win.tree:
            win.tree.node.raise_to_top()

    def render_window_borders(self, render_node):
        render_output = render_node.render_output
        borders = [
            node.as_rect for node in render_node.node.as_tree.children
            if node.type == SceneNodeType.RECT
        ]
        if len(borders) != 6:
            return
        tl, br, top, left, right, bottom = borders

        # First render corners
        program = self.programs['border']
        render_region = self.render_border_setup(
            render_node, program, borders=[tl, br])
        if render_region:
            for rect in render_region.rectangles_as_boxes():
                self.scissor_in_damage_coords(render_output, rect)
                GL.glDrawArrays(GL.GL_TRIANGLE_STRIP, 0, 4)

            GL.glDisableVertexAttribArray(program.attributes['pos'])
            render_region.fini()

        GL.glEnable(GL.GL_BLEND)

        program = self.programs['border_fade']
        GL.glUseProgram(program.program)
        color = rgb(render_node.win.bordercolor[0])
        GL.glUniform4f(program.uniforms['color'], *color)
        GL.glVertexAttribPointer(
            program.attributes['pos'], 2,
            GL.GL_FLOAT, GL.GL_FALSE, 0, self.quad_verts)
        GL.glEnableVertexAttribArray(program.attributes['pos'])

        tr = WlOutput.transform
        transforms = [
            tr.normal,
            tr.transform_270,
            tr.transform_90,
            tr.transform_180
        ]
        for rect, transform in zip([top, left, right, bottom], transforms):
            rn = render_node.child(rect.node)
            if not rn.render_region:
                rn.finalize()
                continue

            box = rn.dst_box
            matrix = Matrix.project_box(
                box, transform, 0.0,
                render_output.wlr_output.transform_matrix)
            gl_matrix = (render_output.projection @ matrix).transpose()
            GL.glUniformMatrix3fv(
                program.uniforms['proj'], 1, GL.GL_FALSE, list(gl_matrix._ptr))
            if transform in [tr.normal, tr.transform_180]:
                GL.glUniform2f(program.uniforms['boxSize'],
                               rn.dst_box.width, rn.dst_box.height)
            else:
                GL.glUniform2f(program.uniforms['boxSize'],
                               rn.dst_box.height, rn.dst_box.width)
            scale = render_node.render_output.wlr_output.scale
            GL.glUniform1f(program.uniforms['fadeFull'], 100 * scale)
            GL.glUniform1f(program.uniforms['fadeLen'], 100 * scale)

            for rect in rn.render_region.rectangles_as_boxes():
                self.scissor_in_damage_coords(render_output, rect)
                GL.glDrawArrays(GL.GL_TRIANGLE_STRIP, 0, 4)

            rn.finalize()

        GL.glDisableVertexAttribArray(program.attributes['pos'])

wl_renderer = MyOtherRenderer
```

![screenshot](https://github.com/qtile/qtile/assets/7871917/083f19a8-54e5-45de-a422-b4f990b2efce)

Subclassing your own renderer requires no extra dependencies, no recompiling, nothing: just code in your config file.  And hopefully you'll agree that writing a scene graph renderer really isn't that hard to do in Python.  Isn't hackability the main point of qtile, and the reason it's written in Python in the first place?  In this case, I'd argue that a freestanding implementation is justified.

Again though, please tell me what you think.  Supporting a custom renderer in qtile (without including one) requires minimal code changes, and no changes in functionality.  Then I can make my desktop look the way I like, and I'll be happy.  But I'm guessing other people will be interested too.

**Reply by justinesmithies** [reply]
Wow that is really really cool !

**Reply by Shinyzenith** [reply]
Wow this is really cool!! Great stuff!

**Reply by jwijenbergh** [reply]
This is very cool, thanks a lot for working on this!

I think the API you made looks good and I can definitely see the benefit. Maybe we can find a way to implement this into qtile. PRs are welcome of course. Maybe we can make it so that by default the wlroots renderer is used with optional scenefx and if people want more control they could plug in their own renderer using a similar API that you have presented here.

Very cool that you managed to make a prototype in such short time. Impressive! The main concern I would have is the fact that a renderer in Python would probably not perform as well as a c renderer, but I have no data to back that up. The OpenGL lib that you use is also using C afaik but that does mean that we have to cross the C FFI barrier more times than when we have an implementation with the default wlroots renderer, or scenefx. *But* as I said having this as an option is really cool, the no recompliation is a big benefit for sure and fits right into the model of qtile

**Reply by QBobWatson** [reply]
Thank you for the post—I appreciate the support. 

> Maybe we can make it so that by default the wlroots renderer is used with optional scenefx and if people want more control they could plug in their own renderer using a similar API that you have presented here.

This occurred to me too. It should be possible to load the wlroots or scenefx renderers with basically the same api—so by default, `wl_renderer = WlrootsRenderer`—give me a few days to work that out.

> Very cool that you managed to make a prototype in such short time.

This kind of code is mostly bookkeeping—much easier in Python than in C.

> The main concern I would have is the fact that a renderer in Python would probably not perform as well as a c renderer

Undoubtedly. But I strongly suspect that enabling any kind of blur effect would have 10x the performance hit either way. And in any case, if performance is your main priority, you’re probably running Sway.

I’m going to work on blur next.  The rendering code should be straightforward, but it’ll be a pain to figure out which surfaces should be part of the window, and which should be above (so the window underneath is blurred). Any advice?

**Reply by WillPower3309** [reply]
> Thank you for the post—I appreciate the support. 
> 
> > Maybe we can make it so that by default the wlroots renderer is used with optional scenefx and if people want more control they could plug in their own renderer using a similar API that you have presented here.
> 
> This occurred to me too. It should be possible to load the wlroots or scenefx renderers with basically the same api—so by default, `wl_renderer = WlrootsRenderer`—give me a few days to work that out.
> 
> > Very cool that you managed to make a prototype in such short time.
> 
> This kind of code is mostly bookkeeping—much easier in Python than in C.
> 
> > The main concern I would have is the fact that a renderer in Python would probably not perform as well as a c renderer
> 
> Undoubtedly. But I strongly suspect that enabling any kind of blur effect would have 10x the performance hit either way. And in any case, if performance is your main priority, you’re probably running Sway.
> 
> I’m going to work on blur next.  The rendering code should be straightforward, but it’ll be a pain to figure out which surfaces should be part of the window, and which should be above (so the window underneath is blurred). Any advice?

Feel free to look into the SwayFX PR that implemented blur as a reference, it's easy to find on our repo. I'd caution against the assumption that it's straightforward, especially in the damage tracking department, however you've clearly shown to be adept at implementing similar features, so I'm sure you'll figure it out!

**Reply by QBobWatson** [reply]
> Feel free to look into the SwayFX PR that implemented blur as a reference, it's easy to find on our repo. I'd caution against the assumption that it's straightforward, especially in the damage tracking department, however you've clearly shown to be adept at implementing similar features, so I'm sure you'll figure it out!

Well I got the blur draw calls working, but I see what you mean about damage tracking.  I really _don't_ want to rerender the blur every time the surface on _top_ is damaged.  It looks like that's what's happening in swayfx and scenefx though, or am I missing something?  How do you get around this?

**Comment by QBobWatson** [comment]
I got blur working now.  Damage tracking also works, although it's not very efficient given the limitations of wlroot's damage ring (which I'm not interested in rewriting).  At blur level 4, any change to a blurred pixel requires a 96x96 box around it to be redrawn!  This is because a pixel can affect a blurred pixel's color up to 48 pixels away.

![screenshot](https://github.com/qtile/qtile/assets/7871917/24afc581-68fe-4f90-a6db-0a3af61e3ed6)

I put the custom renderer in a gist for now: https://gist.github.com/QBobWatson/07e814a2d4eeee22741d228b918c720f
It requires https://github.com/qtile/qtile/pull/4580 and https://github.com/flacjacket/pywlroots/pull/131 to work.

I'll do drop shadows after Thanksgiving.
